content/en/blog/low-hanging-performance/index.md

```diff
@@ -0,0 +1,460 @@
+an assert package.
+
+## Function Inline Expansion
+
+Let's write our version of the famous `assert` function to show how function
+inlining can help readability outside of the tests without sacrificing
+performance.
+
+```go
+func assert(term bool, msg string) {
+    if !term {
+        panic(msg)
+    }
+    ...
+
+func doSomething(p any, b []byte) {
+    assert(p != nil, "interface value cannot be nil")
+    assert(len(b) != 0, "byte slice cannot be empty (or nil)")
+    ...
+    // coninuet with something imporant
+```
+
+By writing the benchmark function for `assert` with Go's testing capabilities,
+you can measure the 'weight' of the function itself. You get the comparison
+point by writing the reference benchmark where you have manually inline-expansed
+the function, i.e. by hand. However, if you aren't interested in the actual
+performance figures but just the information about successful inline expansion
+done by the compiler, you can ask:
+
+```
+go test -c -gcflags=-m=2 <PKG_NAME> 2>&1 | grep 'inlin'
+```
+
+The `-gcflags=-m=2` gives lots of information, but we can filter only those
+lines that contain messages inlining. Depending on the size of the packages
+there can be an overwhelming lot of information where most of them aren't
+related to the task in your hand. You can always filter more.
+
+The `-gcflags` will be your programming buddy in the future. To get more
+information about the flags, run:
+
+```
+go build -gcflags -help
+```
+
+## Memory Allocations
+
+Similarly, as function calls, the memory allocations from the heap are
+expensive. It’s good practice to prevent unnecessary allocations even when the
+programming platform has a garbage collector. With the Go, it’s essential to
+understand the basics of memory management principles Go uses because of memory
+locality, i.e., it has pointers and value types. Many other garbage-collected
+languages have object references, and the [memory
+locality](https://www.youtube.com/watch?v=bmZNaUcwBt4&t=1626s) is hidden from
+the programmer, leading to poor performance, e.g., cache misses.
+
+But nothing comes for free – you need to know what you’re doing. Go’s compiler
+analyzes your code and, without your help, can decide if a variable is *escaping*
+from its scope and needs to be moved from a stack to the heap.
+
+Go’s tools give you extra information about [escape
+analyzes](https://appliedgo.com/blog/how-to-do-escape-analysis). Use the
+`-gcflags=-m=2` again, but `grep` *escape* lines from the output. That will tell
+you exactly what’s going on with the pointers for every function in the current
+compilation.
+
+Usually, when benchmarking Go code, it’s good to understand what’s going on with
+heap allocations. Just add the following argument e.g., your test benchmark
+compilation, and you get the statistics of allocations in the benchmark run.
+
+```
+go test -benchmem -bench=. <PKG_NAME>
+```
+
+The `-benchmem` flag inserts two columns to benchmarking results:
+
+```
```

Comment:
It is a bit confusing that the sources for these functions are shown only after this chapter.